app/jobs/process_digital_object_import_job.rb

```diff
@@ -1,82 +1,149 @@
+      # We're updating data for an existing object
+      digital_object = existing_object_for_update(digital_object_data, user)
     else
-      digital_object = new_object(digital_object_data, user, digital_object_import)
+      digital_object = new_object(digital_object_data['digital_object_type']['string_key'], user, digital_object_import)
     end
 
-    status = assign_data(digital_object, digital_object_data, digital_object_import)
+    result = assign_data(digital_object, digital_object_data, digital_object_import)
 
-    return if status == :requeued
+    if result == :parent_not_found
+      # If the parent wasn't found, wait 30 seconds in case the parent is
+      # currently being processed, or if solr hasn't auto-committed yet,
+      # and then try one more time.
+      sleep 30
+      digital_object_import.digital_object_errors = [] # clear errors from last attempt
+      result = assign_data(digital_object, digital_object_data, digital_object_import)
+    end
 
-    handle_success_or_failure(status, digital_object, digital_object_import)
+    if result == :parent_not_found
+      # The referenced parent object(s) still cannot be found, even after a
+      # wait and retry, so we'll mark this as a failure.
+      digital_object_import.digital_object_errors << "Failed because referenced parent object could not be found."
+      digital_object_import.status = :failure
+      digital_object_import.save!
+      return
+    else
+      handle_success_or_failure(result, digital_object, digital_object_import, HYACINTH['solr_commit_after_each_csv_import_row'])
+    end
   end
 
-  def self.handle_success_or_failure(status, digital_object, digital_object_import)
-    if status == :success && digital_object.save(HYACINTH['solr_commit_after_each_csv_import_row'])
-      digital_object_import.success!
-    else # if status == :failure
-      digital_object_import.digital_object_errors += digital_object.errors.full_messages if digital_object.present?
-      digital_object_import.failure!
-    end
+  def self.existing_object?(digital_object_data)
+    digital_object_data['pid'].present? && DigitalObject::Base.exists?(digital_object_data['pid'])
   end
 
   def self.assign_data(digital_object, digital_object_data, digital_object_import)
-    return :failure unless digital_object.present?
     digital_object.set_digital_object_data(digital_object_data, true)
     :success
   rescue Hyacinth::Exceptions::ParentDigitalObjectNotFoundError => e
     digital_object_import.digital_object_errors << e.message
-    requeue_job(digital_object_import)
-  rescue Hyacinth::Exceptions::NotFoundError, Hyacinth::Exceptions::MalformedControlledTermFieldValue, UriService::Error => e
-    digital_object_import.digital_object_errors << e.message
+    :parent_not_found
+  rescue Exception => e
+    digital_object_import.digital_object_errors << e.message + '<span class="backtrace">' + "Backtrace:\n\t#{e.backtrace.join("\n\t")}" + '</span>'
```

Comment:
I feel like the <span> should be moved into a template of some sort, but that would probably mean an update to how errors are stored. Again, not a deal breaker.

Reply:
I'll make a template method to format this string.  Good idea!